src/imp/schannel.rs

```diff
@@ -559,4 +692,179 @@ mod pem {
             vec![b"-----BEGIN garbage" as &[u8]]
         );
     }
+
+    #[test]
+    fn test_parse_engine_string() {
+        
+        let my_os_str = OsStr::new("user:my:7b78a8e15d5ddfccaa71088ee44606981bb804d7");
+        let my_file_str = OsStr::new(r"file:C:\Microsoft.Autopilot.Security\MachineFunctionCerts\CY2TEAP00013459.CY2Test01.sst"); // have to prefix with "r" for raw string
+        let file_result = parse_engine_string(my_file_str).unwrap();
+        let os_result = parse_engine_string(my_os_str).unwrap();
+
+        // Verify file parse
+        assert!(matches!(file_result, OsProviderParameters::ContextFromFile{..}));
+        match file_result {
+            OsProviderParameters::ContextFromFile {file_path} => {
+                assert_eq!(file_path.as_os_str(), r"C:\Microsoft.Autopilot.Security\MachineFunctionCerts\CY2TEAP00013459.CY2Test01.sst");
+            }
+            _ => {}
+        }
+
+        // Verify store parse
+        assert!(matches!(os_result, OsProviderParameters::ContextFromStore{..}));
+
+        match os_result {
+            OsProviderParameters::ContextFromStore {store_name, is_machine, decoded_hex} => {
+                let expected_hex = hex::decode("7b78a8e15d5ddfccaa71088ee44606981bb804d7").unwrap();
+                assert_eq!(false, is_machine);
+                assert_eq!(store_name, "my".to_string());
+                assert_eq!(expected_hex, decoded_hex)
+            }
+            _ => {}
+        }
+
+        // Test parse errors
+        let my_os_str = OsStr::new("willfail:my:7b78a8e15d5ddfccaa71088ee44606981bb804d7");
+        assert!(parse_engine_string(my_os_str).is_err());
+
+        let my_os_str = OsStr::new("user:invalidstorename:7b78a8e15d5ddfccaa71088ee44606981bb804d7");
+        assert!(parse_engine_string(my_os_str).is_err());
+        
+        let my_os_str = OsStr::new("user:my:thishashistooshort");
```

Comment:
>thishashistooshort [](http://example.com/codeflow?start=44&length=18)

invalid hash characters.

You need another case only using ascii hex characters.